### ISSUE TITLE
Migrate to pac4j 2.x

### ISSUE THREAD

**Issue by millross** [pac4j/vertx-pac4j#91]
Update the vert.x-pac4j project to use pac4j 2.x (milestones initially).

**Comment by ssekaran23** [comment]
Hi Team,

Just wanted to know by when the migration to pac4j 2.x will be completed. I am looking for pac4j 2.x to support global logout.

Regards,
Srikanth

**Comment by millross** [comment]
It will be completed when it's completed. It's a spare time project for me. The release date depends simply on how long the work takes and how much time I have to devote to it.

You're welcome to contribute.

**Comment by millross** [comment]
Also note that the release will lag the 2.0 release of pac4j by some time. The 2.0 release of pac4j has not yet occurred.

**Comment by millross** [comment]
Closed in error.

You should note, however, that completion of this issue does not give you the central logout functionality, which is dealt with by issue #68 . The first, but necessary step is a port of existing functionality to pac4j 2 (and hopefully vert.x 3.4). The second step will be supporting the key new features, each of which has been given its own issue in this repository. We will not issue a release until those new features, including central logout, are completed and we have the demo up to date.

**Comment by leleuj** [comment]
At least, the 2.0.0-RC1 release is available for early adopters and there won't be many API changes until the final release.

**Comment by millross** [comment]
Now up to date with the latest pac4j 2.x snapshot. When the next milestone release occurs I'll update to work against that.

**Comment by leleuj** [comment]
Thanks

**Comment by millross** [comment]
As a general progress report, all is done aside from centralLogout, which is work in progress. I'm seeing a test failure at the moment, and I think it's because of a subtlety around the implementation of DefaultLogoutLogic (writing "" to the WebContext initially, with a 200 status, then switching to a 302 status). I think I have a workaround, but if not I'll be submitting a PR for a change to DefaultLogoutLogic to delay side effects on the WebContext to end of processing logout, rather than issuing two sets of effects.

**Comment by millross** [comment]
Once this is done, I will still have to bring the demo up to date, but hopefully that will not prove too painful.

**Comment by millross** [comment]
centralLogout is now supported in snapshot. I'll start looking at updating the demo next :)

**Comment by millross** [comment]
Build problem due to recent compatibility breaks in pac4j snapshot. I will resolve when I have time, but this may not be before Tuesday 28/03/17 - until then the centralLogout functionality will remain unavailable in vertx-pac4j

**Comment by millross** [comment]
We've had to rollback from snapshot to build against pac4j 2.0.0-M1. Some of the changes in 2.0.0-M2 (and on SNAPSHOT over the past week) break vertx-pac4j.

So centralLogout is now supported, but for the time being if you wish to use it, you must use pac4j 2.0.0-M1.

**Comment by millross** [comment]
Sorry, I referred to M1 and M2 when I should have referred to RC1 and RC2. However, the bottom line is still the same. vertx-pac4j  cannot pass tests against RC-2 due to issue #871 on pac4j-core. So we currently build against RC-1. When 871 is fixed in core, we'll reinstate building against pac4j 2.0.0-SNAPSHOT. However, there will not be a version of vertx-pac4j which supports RC2

**Comment by leleuj** [comment]
Copy that. Waiting for your PR to fix pac4j#871 for a merge...

**Comment by millross** [comment]
@ssekaran23 It looks as though centralLogout is now being implemented on a per-client basis. The current default implementation for CAS is not vertx-compatible. I suspect the other ones as they arrive also won't be vertx-compatible. So you may still have longer to wait for the ability to run a central logout with your chosen technology. I strongly suggest that, for your chosen client, you investigate how central logout should be implemented and submit a vertx-compatible PR. I will be working on the CAS version but probably not for the initial release of vertx-pac4j, as we will also be using that implementation to feed back into a nicer central logout API (minimising the number of components which have to be modified) for a new central logout implementation.
